imagepuller/internal/service/verify.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2025 Edgeless Systems GmbH
+// SPDX-License-Identifier: BUSL-1.1
+
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+)
+
+func (s *ImagePullerService) getAndVerifyImage(ctx context.Context, log *slog.Logger, imageURL string) (v1.Image, error) {
+	ref, err := name.ParseReference(imageURL)
+	if err != nil {
+		return nil, fmt.Errorf("parsing the image URL as a reference: %w", err)
+	}
+
+	digestExpectedRaw, err := name.NewDigest(imageURL)
+	if err != nil {
+		return nil, fmt.Errorf("parsing image digest: %w", err)
+	}
+	digestExpected := digestExpectedRaw.DigestStr()
+
+	tr := transport.NewRetry(remote.DefaultTransport)
+
+	remoteImgIndex, err := remote.Index(ref, remote.WithContext(ctx), remote.WithTransport(tr))
+	if err == nil {
+		log.Info("Received manifest list")
+
+		digestGot, err := remoteImgIndex.Digest()
+		if err != nil {
+			return nil, fmt.Errorf("obtaining actual image index digest: %w", err)
+		}
+		if digestGot.String() != digestExpected {
+			return nil, fmt.Errorf("validating image index: expected digest '%s', got digest '%s'", digestExpected, digestGot)
+		}
+		log.Info("Validated image index digest")
+
+		manifest, err := remoteImgIndex.IndexManifest()
+		if err != nil {
+			return nil, fmt.Errorf("obtaining index manifest: %w", err)
+		}
+
+		var digestFound string
+		for _, m := range manifest.Manifests {
+			if m.Platform.String() == "linux/amd64" {
+				digestFound = m.Digest.String()
+				break
+			}
+		}
+
+		if digestFound == "" {
+			return nil, fmt.Errorf("obtaining image digest for linux/amd64: platform missing from image index")
+		}
+
+		ImageURLLinux := strings.Replace(imageURL, digestExpected, digestFound, 1)
+		log.Info("Obtained actual image digest", "image_url_linux", ImageURLLinux)
+		return s.getAndVerifyImage(ctx, log, ImageURLLinux)
+	}
+
+	remoteImg, err := remote.Image(ref, remote.WithContext(ctx), remote.WithTransport(tr))
+	if errors.Is(err, context.DeadlineExceeded) {
+		return nil, fmt.Errorf("pull aborted (deadline exceeded): %w", err)
+	} else if err != nil {
+		return nil, fmt.Errorf("accessing the remote image URL: %w", err)
+	}
```

Comment:
You're doing the recursion again. Instead, you could simply do

```suggestion
		digestExpected = digestFound
		
		remoteImg, err = remoteImgIndex.Image(digestFound) // actually, arg needs to be `m.Digest`
	} else {
		remoteImg, err = remote.Image(ref, remote.WithContext(ctx), remote.WithTransport(tr))
		if err != nil {
			return nil, fmt.Errorf("accessing the remote image URL: %w", err)
		}
	}
```

Reply:
Ah! Wasn't aware about remoteImgIndex.Index, thank you!